### ISSUE TITLE
Training Stopped Prematurely

### ISSUE THREAD

**Issue by Voldy1998** [AntonMu/TrainYourOwnYOLO#78]
Before filing a report consider the following two questions:

### Have you checked the [troubleshooting] Yes (https://github.com/AntonMu/TrainYourOwnYOLO#troubleshooting) section?

Once you are familiar with the code, you're welcome to modify it. Please only continue to file a bug report if you encounter an issue with the provided code and after having followed the instructions.

If you have followed the instructions exactly, couldn't solve your problem with the provided troubleshooting tips and would still like to file a bug or make a feature requests please follow the steps below.

1. It must be a bug, a feature request, or a significant problem with the documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.

------------------------

### System information
- **What is the top-level directory of the model you are using**: I didn't understand  C:\Users\anjana ouseph\Desktop\Train\custom\TrainYourOwnYOLO\2_Training> Is this what you mean? i have created a virtual env.
- **Have I written custom code (as opposed to using a stock example script provided in the repo)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **TensorFlow version (use command below)**: The one in requirements.txt
- **CUDA/cuDNN version**: I don't have CUDA installed
- **GPU model and memory**: NVIDIA GEforce 940MX,4GB graphics card  RAM 12GB 
- **Exact command to reproduce**: python detector.py

You can obtain the TensorFlow version with

`python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`

### Describe the problem
My training stopped at 78/102 epochs. It stopped prematurely. when i tried running the detector using the saved model i got okayish results but no that great. What is the issue here? I am running it on CPU as i don't have cuda installed. It's been training for one whole day.
![yolo](https://user-images.githubusercontent.com/40512980/73270927-b3fa0c80-4205-11ea-81b7-a980fa2bb72a.PNG)

### Source code / logs
90/90 [==============================] - 1881s 21s/step - loss: 13.1176 - val_loss: 16.9962
Epoch 00078: early stopping

**Comment by AntonMu** [comment]
Hi @Voldy1998 - thanks for completing the issue template. 

It looks like everything worked as expected from a software side. Early stopping is a built-in feature that automatically stops training if the validation loss is no longer improving - meaning that your model isn't getting any better. 

But I can see that it is relatively high and thus probably not giving you the results you expect. At this point you would need to start thinking about your experiment design. Are you using enough training images? Is there enough similarity between objects? Are your objects similar enough to the pre-trained weights (which I believe are trained in ImageNet). Any other potential issues?

Also, to speed up training you can probably reduce the number of epochs to 10 or 20 to get a better idea of performance before training a full model. Good luck!

**Comment by Voldy1998** [comment]
Thank you so much for the quick and detailed reply @AntonMu . This tutorial is the best i have seen so far. 
1). I would like to know whether annotations can affect the training process too? For example I am attaching an image which shows how i have annotated an image of my dataset consisting of cars. In it I have tried to draw a bounding box over cars that appear in the background too (the ones which appear only partially and are not much clear), is that necessary?
2).  How to reduce the epochs to 10 or 20? which part of the code does that?
When i trained now it had 51 epochs after completing that, it started again with 102 epochs and stopped at 78th epoch.
3). Also do you know how to install cuda and associated things necessary to train on a GPU in windows?
![annontations](https://user-images.githubusercontent.com/40512980/73291933-2a5b3680-4227-11ea-848e-3d067282de70.PNG)

**Comment by kidboy-man** [comment]
@Voldy1998 
1) Based on my experience using yolo, it is necessary to label all the classes that appear in the images. This will make the model more versatile to recognize an object. Also, if we don't label the object that under the same class in an image,  it will confuse the model.
2) You can do it by executing command `python Train_YOLO.py --epochs 20`.
3) If you have trouble installing CUDA requierements, you can run the training on google colab, it also provides free Tesla GPU

**Comment by kidboy-man** [comment]
Hi @AntonMu 
I also have a question regarding this issue. Is it possible to stop training and then resuming it from the last epoch or checkpoint? because often my runtime in google colab exceeded the limit of 12 hours uses, when I redo training it always starts from epoch 0 again. Any suggestion on this? Thank you

**Comment by robisen1** [comment]
> Hi @AntonMu
> I also have a question regarding this issue. Is it possible to stop training and then resuming it from the last epoch or checkpoint? because often my runtime in google colab exceeded the limit of 12 hours uses, when I redo training it always starts from epoch 0 again. Any suggestion on this? Thank you

-----------------------------------------------------------------------------------------------

Could you try to transfer learning? Useless data and save the model. Then add more data and add to old model etc?

**Comment by AntonMu** [comment]
Hi - yes, you can resume training from a previous checkpoint. Just pass it in as an argument in the train function [here](https://github.com/AntonMu/TrainYourOwnYOLO/blob/6cb42c98d160bcc865a09bbb92615019df0278d8/2_Training/Train_YOLO.py#L109).

**Comment by bukolastella** [comment]
> Hi - yes, you can resume training from a previous checkpoint. Just pass it in as an argument in the train function [here](https://github.com/AntonMu/TrainYourOwnYOLO/blob/6cb42c98d160bcc865a09bbb92615019df0278d8/2_Training/Train_YOLO.py#L109).



> Hi - yes, you can resume training from a previous checkpoint. Just pass it in as an argument in the train function [here](https://github.com/AntonMu/TrainYourOwnYOLO/blob/6cb42c98d160bcc865a09bbb92615019df0278d8/2_Training/Train_YOLO.py#L109).

Anyone have the code snippet or example to do this
